apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+                        {
+                            // Retry for 600 seconds.
+                            RetryTemporaryFailures(idsToRetry, acksToRetry => queue.Locked(() => queue.Requeue(acksToRetry)), Enumerable.Empty<TimedId>(), null, 600);
+                        }
+                    }
+
+                    if (ackError.TryGetAcknowledgementException(out var exception))
+                    {
+                        if (hasExtendIds)
+                        {
+                            // Don't bubble permanent failures for lease extends as they are not user initiated.
+                            // Just log them.
+                            // TODO: Log the permanent failures.
+                            return;
+                        }
+
+                        // For acks/nacks, bubble up the exception.
+                        throw exception.FlattenIfPossible();
+                    }
+                }
+
+                // Retries the ackIds/nackIds with temporary failures for Acknowledge RPC or extendIds with temporary failures for ModifyAckDeadline RPC.
+                // This method checks if the id exists in _retryableIds dictionary and has not been tried for more than the specified timeout.
+                // If so, it retries those ids by adding them to the appropriate queue and calls the StartPush() method.
+                // For acks and nacks, ackOrnackIds must be non-null. For extends, extendIds must be non-null.
+                // For acks and nacks, ackActionToRetry must be non-null and specify the right queue (_ackQueue or _nackQueue).
+                // For extends, extendActionToRetry must be non-null and specify the right queue,i.e, _extendQueue.
+                void RetryTemporaryFailures(
+                    IEnumerable<string> ackOrNackIds,
+                    Action<IEnumerable<string>> ackActionToRetry,
+                    IEnumerable<TimedId> extendIds,
+                    Action<IEnumerable<TimedId>> extendActionToRetry = default,
+                    int timeoutInSeconds = 600)
+                {
+                    // TemporaryFailures may be empty but will never be null, so null check is not needed.
+                    // By design, only items in temporaryFailures that exist in _retyrableIds will be retried.
+                    // If id is not in _retryableIds dictionary, it is either already succeeded or not retryable.
+                    // Failure may be non-retryable due to exceeded time duration or that failure may have become permanent.
+
+                    bool hasAcksOrNacks = ackOrNackIds.Any();
+                    bool hasExtends = extendIds.Any();
+                    if (!hasAcksOrNacks && !hasExtends)
+                    {
+                        return;
+                    }
+
+                    var idsToRetry = hasAcksOrNacks ? ackOrNackIds : extendIds.Select(j => j.Id);
+                    var retryableErrorIds = new ConcurrentBag<string>();
```

Comment:
I don't think this needs to be a ConcurrentBag - I suspect it could just be a `List<string>`